### PR TITLE
Wrapping up GEMINI development

### PR DESCRIPTION
**Pull request by haitamlaarabi**
+ Integrating Advanced Public SPMC
+ Integrating Ridehail SPMC
+ Updating ScaleUpCharging
+ Updating HELICS deployment
+ Refactoring

### CODE REVIEW

**Comment by haitamlaarabi** [comment]
Test Now!

**Comment by haitamlaarabi** [comment]
Test now!

File: aws/src/main/python/beam_lambda/lambda_function.py [hunk 1]
```diff
@@ -105,6 +104,28 @@
...
+            echo "broker started"
+            sleep 5s
+            python beam_to_pydss_federate.py > output_${now}_federate.log &
+            echo "federate started"
+            sleep 5s
+            helics_recorder beam_recorder.txt --output=recording_output.txt > output_${now}_recorder.log &
+            echo "recorder started"
+            sleep 5s
+            cd -
+      path: /home/ubuntu/install-and-run-helics-scripts.sh
```
Comment (JustinPihony) [inline aws/src/main/python/beam_lambda/lambda_function.py:128]: Does anything call this?

File: src/main/resources/beam-template.conf [hunk 2]
```diff
@@ -391,7 +391,7 @@ beam.agentsim.chargingNetworkManager {
     intLogLevel = "int | 1"
     beamFederatePrefix = "String | BEAM_FED"
     beamFederatePublication = "String | CHARGING_VEHICLES"
-    spmFederatePrefix = "String | SPM_FED_TAZ"
+    spmFederatePrefix = "String | SPM_FED"
```
Comment (JustinPihony) [inline src/main/resources/beam-template.conf:399]: You need to regenerate the code - I do not see that checked in.

File: src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala [hunk 3]
```diff
@@ -49,11 +50,11 @@ class ChargingFunctions(
...
     * @return
     */
-  def ifRideHailCurrentlyOnShiftThenFastChargingOnly(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
+  private def ifRideHailCurrentlyOnShiftThenFastChargingOnly(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
     zone.chargingPointType.forall(chargingPointType =>
       if (
-        inquiry.reservedFor.managerType == VehicleManager.TypeEnum.RideHail || inquiry.beamVehicle
-          .exists(v => v.isRideHail && (inquiry.parkingDuration <= 3600 || v.isCAV))
+        inquiry.parkingDuration <= 3600 && (inquiry.reservedFor.managerType == VehicleManager.TypeEnum.RideHail || inquiry.beamVehicle
+          .exists(_.isRideHail))
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala]: This lost the CAV check - was that on purpose?

**Review by JustinPihony — CHANGES_REQUESTED** [review]
I was light on the R and Python review. Requesting change because of the missing BeamConfig file

File: src/main/python/gemini/cosimulation/nrel_spmc_controller/rudimentary_spmc_v5.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import math
+import numpy as np
+
+#=========================================================
+#      SPM control algorithm starts here
+#      objective: minimize peak power
+#=========================================================     
+class SPM_Control():
+    def __init__(self, time_step_mins, max_power_evse, min_power_evse):
+        self.time_step_mins = 1                   # time step: 1 minutes
```
Comment (HoneyTauOverTwo) [inline src/main/python/gemini/cosimulation/nrel_spmc_controller/rudimentary_spmc_v5.py]: This is ignoring the parameter time_step_mins. Is this intentional?
(just saw spmc_v3.py version of this class uses it, this looks intentional)

**Comment by HoneyTauOverTwo** [comment]
test!

File: src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala [hunk 5]
```diff
@@ -119,7 +125,6 @@ object ParkingZone extends LazyLogging {
       geoId,
       parkingType,
       VehicleManager.AnyManager,
-      Some(SitePowerManager.createId(DefaultParkingZoneId.toString)),
```
Comment (HoneyTauOverTwo) [inline src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala:122]: PowerManagerSpec.scala#L45, SitePowerManager.scala#L95 and ZonalParkingManagerSpec.scala#L624 have to be modified to reflect this change.
(the main change is on line 132 actually)

**Review by HoneyTauOverTwo — CHANGES_REQUESTED** [review]
As pointed out by @JustinPihony, BeamConfig.scala needs to be generated again since beam-template.conf was modified.

Also, there are some tests that need updating since the field ``siteIdMaybe`` was removed from ``ParkingZone.scala``

**Comment by nikolayilyin** [comment]
test!

File: src/main/resources/beam-template.conf (hunk 2 above)
Comment (haitamlaarabi) [reply src/main/resources/beam-template.conf:399]: Done

File: src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala (hunk 3 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala]: I have updated the whole section

File: src/main/python/gemini/cosimulation/nrel_spmc_controller/rudimentary_spmc_v5.py (hunk 4 above)
Comment (haitamlaarabi) [reply src/main/python/gemini/cosimulation/nrel_spmc_controller/rudimentary_spmc_v5.py]: fixed

File: src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala (hunk 5 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala:122]: Done

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

File: src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+
+    val vehicleIds = findAllElectricVehicles(events).map(id => id._1.toString)
+    vehicleIds.size shouldEqual 50 withClue ", expecting 50 electric vehicles."
+
+    val vehiclesCharged = filterEvents(
+      events,
+      ("type", a => a.equals("ChargingPlugInEvent"))
+    ).map(e => e.getAttributes.get("vehicle")).distinct
+
+    vehiclesCharged.size should be > 0 withClue ", every single vehicle should had charged at least once."
```
Comment (HoneyTauOverTwo) [inline src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala]: Originally

`vehiclesCharged.size shouldEqual 50 withClue ", every single vehicle should had charged at least once."`

I guess this test was failing to you and you changed it to `> 0`, is that it?
In this case, the clue text should be changed to match the test.

I also might have to investigate why the original check is not passing anymore, originally, I tried to make sure these tests were not flaky.

File: src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+
+    val (matsimConfig, _, _) = runBeamWithConfig(config)
+
+    val events = EventReader.fromXmlFile(
+      EventReader.getEventsFilePath(matsimConfig, "events", "xml").getAbsolutePath
+    )
+
+    val centerTAZs = List("10", "11")
+    val borderTAZs = List("20", "21", "30", "31")
+
```
Comment (HoneyTauOverTwo) [inline src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala:218]: ~~There used to be a line like this here and other places:~~

~~`val unsuitableChargersTAZs = List("18", "19", "28", "29", "38", "39")`~~

~~which was used to check whether vehicles were plugging into chargers without sufficient power rating. What was the reason for removing this? Was it failing or this just does not make sense to check?~~

~~I have added this check to cover a change I made to how charging stalls are filtered: https://github.com/LBNL-UCB-STI/beam/pull/3727/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR126~~

I just saw you changed how that filter works and now it does not depend on charger power, just it's capability. So the test changes related to this make sense.

**Review by HoneyTauOverTwo — CHANGES_REQUESTED** [review]
All these files:

`test/test-resources/ElectricVehicleChargingBehaviorTestData/...`
`    taz-centers.csv`
`    taz-parking-destination-only.csv`
`    taz-parking-enroute-only-free.csv`
`    taz-parking-enroute-only-mixed-prices.csv`
`    taz-parking-ride-hail.csv`
`    taz-parking-ride-hail-charging-time-cost.csv`
`    taz-parking-ride-hail-driving-time-cost.csv`

Should remove every single entry that has the TAZs `18, 19, 28, 29, 38, 39` since they are not used anymore on `ElectricVehicleChargingBehaviorTest.scala` and since `ChargingFunctions.scala` was also changed, now these chargers are actually valid, which I think is the reason why `ElectricVehicleChargingBehaviorTest` is failing:

`27 was not greater than 41, agents more likely to enroute charge in center TAZs than in border ones.`

Line 120 on `ElectricVehicleChargingBehaviorTest.scala` should also change to reflect these changes (I couldn't add a comment directly there):

```
    TAZ MAP:
                     ---------- 20, 22 ------------ 21, 23 ----------
                     |      167692.88,2213.6    168807.15,2213.6    |
             home(8) ---------- 10, 12 ------------ 11, 13 -- -------- work(9)
    166021.46,1106.8 |      167692.88,1106.8    168807.15,1106.8    | 169921.40,1106.8
                     ---------- 30, 32 ------------ 31, 33 ----------
                            167692.88,5.53      168807.15,5.53
```

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
> All these files:
> 
> `test/test-resources/ElectricVehicleChargingBehaviorTestData/...` ` taz-centers.csv` ` taz-parking-destination-only.csv` ` taz-parking-enroute-only-free.csv` ` taz-parking-enroute-only-mixed-prices.csv` ` taz-parking-ride-hail.csv` ` taz-parking-ride-hail-charging-time-cost.csv` ` taz-parking-ride-hail-driving-time-cost.csv`
> 
> Should remove every single entry that has the TAZs `18, 19, 28, 29, 38, 39` since they are not used anymore on `ElectricVehicleChargingBehaviorTest.scala` and since `ChargingFunctions.scala` was also changed, now these chargers are actually valid, which I think is the reason why `ElectricVehicleChargingBehaviorTest` is failing:
> 
> `27 was not greater than 41, agents more likely to enroute charge in center TAZs than in border ones.`
> 
> Line 120 on `ElectricVehicleChargingBehaviorTest.scala` should also change to reflect these changes (I couldn't add a comment directly there):
> 
> ```
>     TAZ MAP:
>                      ---------- 20, 22 ------------ 21, 23 ----------
>                      |      167692.88,2213.6    168807.15,2213.6    |
>              home(8) ---------- 10, 12 ------------ 11, 13 -- -------- work(9)
>     166021.46,1106.8 |      167692.88,1106.8    168807.15,1106.8    | 169921.40,1106.8
>                      ---------- 30, 32 ------------ 31, 33 ----------
>                             167692.88,5.53      168807.15,5.53
> ```

I updated those files

**Comment by HoneyTauOverTwo** [comment]
test!

**Review by HoneyTauOverTwo — APPROVED** [review]
Looks good to me.

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by HoneyTauOverTwo** [comment]
The issue came back with the test getting stuck

```
18:26:52 *** Test still running after 46 minutes, 22 seconds: suite name: ElectricVehicleChargingBehaviorTest, test name: Ride Hail Electric vehicles should pick chargers choosing smaller DrivingTimeCost.. 
18:26:56 21:26:58.839 [ClusterSystem-akka.actor.default-dispatcher-16] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
18:26:58 21:27:08.859 [ClusterSystem-akka.actor.default-dispatcher-15] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
18:27:08 21:27:18.879 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer
```

We should take care of this before merging. 2 out of 3 attempts got stuck with this.

File: src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala [hunk 8]
```diff
@@ -412,34 +444,34 @@ class ElectricVehicleChargingBehaviorTest
     val closestTAZsCAV = List("12")
```
Comment (HoneyTauOverTwo) [inline src/test/scala/beam/agentsim/agents/ElectricVehicleChargingBehaviorTest.scala]: The test `"Ride Hail Electric vehicles" should "pick chargers choosing smaller DrivingTimeCost."` is sometimes getting stuck on CI, this should be fixed before merging. I'll investigate.

**Review by HoneyTauOverTwo — CHANGES_REQUESTED** [review]

**Comment by HoneyTauOverTwo** [comment]
test!

this one passed: 2 fails out of 4 tries

**Comment by HoneyTauOverTwo** [comment]
test!

failed but with flaky test:
`beam.sim.BeamWarmStartRunSpec.Beam WarmStart must run beamville scenario for two iterations with warmstart with normal and fake skims`

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by haitamlaarabi** [comment]
Test now!

**Review by JustinPihony — APPROVED** [review]

**Comment by HoneyTauOverTwo** [comment]
Latest test got stuck as well. I could not reproduce this issue locally.

I ran the test getting stuck (`Ride Hail Electric vehicles should pick chargers choosing smaller DrivingTimeCost.`) 10 times and had no issues.

I also ran the hole class tests another 7 times and did not encounter any issues (I have seen tests behave a bit different in the past when running them separatelly compared to running the hole class).

Although the tests don't get stuck, I did notice the ChargingNetwork error message does show up locally, but it is only once per test (when it does show):

`ERROR b.a.i.ChargingNetworkManager - unhandled message: BeamAgentSchedulerTimer`

I also notice a similar error message with more info:

`ERROR b.a.agents.ridehail.RideHailAgent - unhandled event: Event(BeamAgentSchedulerTimer,RideHailAgentData(rideHailVehicle-07-L5@GlobalRHM (RH_BEV_L5,8.785613000000001km),Vector(rideHailVehicle-07-L5@GlobalRHM),,0,List(),None)) in state [ Idle ] - vehicle( rideHailVehicle-07-L5@GlobalRHM )`

These detailed errors always refer to fully automated vehicles, but I am unsure if they are the cause for the other error message which seems to be the cause for the test getting stuck.

I will temporarily change the stuck test so it does not have any ride hail vehicle with human driver and run CI a few times to see if it still gets stuck.

**Comment by HoneyTauOverTwo** [comment]
test!

This one passed and the console output ~~does not have any mention of "unhandled message"~~. (actually does, forgot to look at the full log)

**Comment by HoneyTauOverTwo** [comment]
test!

Same as before.

**Comment by JustinPihony** [comment]
I've questioned the changes at [this point](https://github.com/LBNL-UCB-STI/beam/pull/3729/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR49) and it seems plausible it could be causing this?

**Comment by HoneyTauOverTwo** [comment]
> I've questioned the changes at [this point](https://github.com/LBNL-UCB-STI/beam/pull/3729/files#diff-71f5a8e60ec0e9bf216515b48833380006c900ed5dc02a1b6b22ebb5d33c5a8dR49) and it seems plausible it could be causing this?

You're right on the money! I discovered a bit ago that vehicles are trying to charge on parking spots with no charging capability. 
The changes made to `hasValidChargingCapability` are causing this. I am pushing changes in a bit.

**Comment by HoneyTauOverTwo** [comment]
@JustinPihony , @haitamlaarabi I am still unsure if the [current version of `hasValidChargingCapability`](https://github.com/LBNL-UCB-STI/beam/blob/f39d1c08b2e21cf6fd34400e6d276cc7ac3c25ee/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L124) is correct. The current one seems to imply that an EV cannot, in any circumstance, park somewhere where there is not a charger.

[The version present on develop branch](https://github.com/LBNL-UCB-STI/beam/blob/develop/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L126) will only filter out a parking zone if the inquiry is related to an activity that requires charging.

I suggest adding the `verifyCharger` check back into it if we are to allow EVs to park somewhere without a charger:

```
    private def hasValidChargingCapability(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
    val verifyCharger = inquiry.beamVehicle.isDefined &&
      inquiry.beamVehicle.get.beamVehicleType.chargingCapability.isDefined && (
        inquiry.searchMode == ParkingSearchMode.EnRouteCharging ||
        inquiry.parkingActivityType == Charge ||
        inquiry.parkingActivityType == EnRoute
      )
    if (!verifyCharger) {
      return true
    }
    zone.chargingPointType match {
      case Some(chargingPointType) =>
        inquiry.beamVehicle.forall {
          case beamVehicle if beamVehicle.isEV =>
            val vehicleHasDCFasChargingCapability =
              beamVehicle.beamVehicleType.chargingCapability.forall(ChargingPointType.isFastCharger)
            val stationHasDCFasChargingCapability = ChargingPointType.isFastCharger(chargingPointType)
            val hasInvalidChargingCapability = !vehicleHasDCFasChargingCapability && stationHasDCFasChargingCapability
            !hasInvalidChargingCapability
          case _ => false
        }
      case _ =>
        false
    }
  }
```

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by HoneyTauOverTwo** [comment]
I just found out that a similar issue exists on other filters as well. `ifRideHailCurrentlyOnShiftThenFastChargingOnly` for instance:

image

If the method consists only of this kind of statement `zone.chargingPointType.forall` when `chargingPointType` is `None`, the block inside will not be executed at all, and the result will be treated as `true`, which is the opposite of what we want in this case.

I have pushed changes that instead of using `zone.chargingPointType.forall(chargingPointType => [...] ChargingPointType.isFastCharger(chargingPointType)` uses `zone.chargingPointType.exists(ChargingPointType.isFastCharger)` which returns false when the `Option` is `None`.

**Comment by HoneyTauOverTwo** [comment]
test!

failed with `beam.integration.EnrouteChargingSpec.Running a single person car-only scenario should do enroute upon not enough charging`. [EnrouteChargingSpec#L147](https://github.com/LBNL-UCB-STI/beam/blob/hl/integrating-ridehail-spmc/src/test/scala/beam/integration/EnrouteChargingSpec.scala#L147)

It is working locally.

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by HoneyTauOverTwo** [comment]
Previous CI build got stuck on the same test `"Ride Hail Electric vehicles" should "pick chargers choosing smaller DrivingTimeCost."`

It appears like the fixes made to prevent RH vehicles to try to charge on parking zones with no charger was not the cause of this issue.

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by HoneyTauOverTwo** [comment]
> @JustinPihony , @haitamlaarabi I am still unsure if the [current version of `hasValidChargingCapability`](https://github.com/LBNL-UCB-STI/beam/blob/f39d1c08b2e21cf6fd34400e6d276cc7ac3c25ee/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L124) is correct. The current one seems to imply that an EV cannot, in any circumstance, park somewhere where there is not a charger.
> 
> [The version present on develop branch](https://github.com/LBNL-UCB-STI/beam/blob/develop/src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala#L126) will only filter out a parking zone if the inquiry is related to an activity that requires charging.
> 
> I suggest adding the `verifyCharger` check back into it if we are to allow EVs to park somewhere without a charger:
> 
> ```
>     private def hasValidChargingCapability(zone: ParkingZone, inquiry: ParkingInquiry): Boolean = {
>     val verifyCharger = inquiry.beamVehicle.isDefined &&
>       inquiry.beamVehicle.get.beamVehicleType.chargingCapability.isDefined && (
>         inquiry.searchMode == ParkingSearchMode.EnRouteCharging ||
>         inquiry.parkingActivityType == Charge ||
>         inquiry.parkingActivityType == EnRoute
>       )
>     if (!verifyCharger) {
>       return true
>     }
>     zone.chargingPointType match {
>       case Some(chargingPointType) =>
>         inquiry.beamVehicle.forall {
>           case beamVehicle if beamVehicle.isEV =>
>             val vehicleHasDCFasChargingCapability =
>               beamVehicle.beamVehicleType.chargingCapability.forall(ChargingPointType.isFastCharger)
>             val stationHasDCFasChargingCapability = ChargingPointType.isFastCharger(chargingPointType)
>             val hasInvalidChargingCapability = !vehicleHasDCFasChargingCapability && stationHasDCFasChargingCapability
>             !hasInvalidChargingCapability
>           case _ => false
>         }
>       case _ =>
>         false
>     }
>   }
> ```

For some reason, this change fixes the failing test locally `beam.integration.EnrouteChargingSpec.Running a single person car-only scenario should do enroute upon not enough charging`

I'll push this change and run CI again. @haitamlaarabi, please review the quoted message and let me know if you agree with it.

**Comment by HoneyTauOverTwo** [comment]
test!

passed (test getting stuck skipped)

**Comment by HoneyTauOverTwo** [comment]
test!

(same as before)

**Comment by HoneyTauOverTwo** [comment]
test!

passed, including the test previously getting stuck. Will run CI another 2 times at least to make sure.

**Comment by HoneyTauOverTwo** [comment]
test!

**Comment by HoneyTauOverTwo** [comment]
test!

CI wasn't working.

**Comment by HoneyTauOverTwo** [comment]
test!

CI wasn't working.

**Comment by HoneyTauOverTwo** [comment]
test!

This passed, testing again to see it if was a fluke.

**Comment by HoneyTauOverTwo** [comment]
test!

Failed with test stuck.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
test!

Still gets stuck without RH with human drivers, so it is most likely something related to L5 RH or both.

**Comment by HoneyTauOverTwo** [comment]
test!

fmt failed.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
test!

Passed.

**Comment by HoneyTauOverTwo** [comment]
I was able to reproduce the stuck issue locally as well, and it is definitely related to parallelism.

Also, from the CI results, it seems like the issue is only related to L5 automated RH electric vehicles, so RH electric vehicles with human drivers is not the source of the issue.

**Comment by HoneyTauOverTwo** [comment]
test!

**Review by HoneyTauOverTwo — APPROVED** [review]
Approving since AV RH issue was proven to not be caused by these changes.
This issue will continue to be investigated on #3766.

**Pull request merged**